test/caPutJsonLogTest.cpp

```diff
@@ -649,14 +669,20 @@ void testDbf<dbr_string_t>(const char *pv, chtype type,
 
         // New Value check
         for (size_t i = 0; i < json.newVal.size(); i++) {
-            testOk(!json.newVal.at(i).compare(value2 + i*MAX_STRING_SIZE),
-                    "%s - %s", testPrefix, "New value check");
+            char *cval = value2 + i*MAX_STRING_SIZE;
+            std::string sval(cval, std::min<size_t>(MAX_STRING_SIZE, strlen(cval)));
+            testOk(!json.newVal.at(i).compare(sval),
+                   "%s - %s - elem %lu exp '%s' act '%s'", testPrefix, "New value check 2",
+                   i, sval.c_str(), json.newVal.at(i).c_str());
```

Comment:
I'm a bit surprised at this. I assume that this is what you mentioned about comparisons between strings with data copied after the null byte?

Reply:
EPICS strings have 40 characters in 40 bytes, so they might have a terminating NULL or not.
std::string allows NULL characters embedded in strings.

std::string(const char *) would overread if the EPICS string is 40 characters long.
std::string(const char *, size_t) would copy junk values after a NULL.

Reply:
The other problem is - as you wrote - when an array of `DBR_STRING` contains 40 character "full size" strings, which would have the "C string" constructor of `std::string` concatenate elements.

The correct way is "until the NULL byte or the max. length, whatever happens first".